pkg/cmd/experimental/diagnostics/diagnostics.go

```diff
@@ -0,0 +1,198 @@
+package cmd
+
+import (
+	"fmt"
+	"github.com/openshift/origin/pkg/cmd/experimental/diagnostics/options"
+	"github.com/openshift/origin/pkg/cmd/server/start"
+	"github.com/openshift/origin/pkg/cmd/templates"
+	osclientcmd "github.com/openshift/origin/pkg/cmd/util/clientcmd"
+	"github.com/spf13/cobra"
+	"io"
+	//"github.com/openshift/origin/pkg/cmd/cli/cmd"
+	//"github.com/openshift/origin/pkg/cmd/cli/config"
+	//"github.com/openshift/origin/pkg/diagnostics/types"
+	//"github.com/openshift/origin/pkg/diagnostics/log"
+	"github.com/openshift/origin/pkg/diagnostics/run"
+)
+
+const longAllDescription = `
+OpenShift Diagnostics
+
+This command helps you understand and troubleshoot OpenShift. It is
+intended to be run from the same context as an OpenShift client or running
+master / node in order to troubleshoot from the perspective of each.
+
+    $ %[1]s
+
+If run without flags or subcommands, it will check for config files for
+client, master, and node, and if found, use them for troubleshooting
+those components. If master/node config files are not found, the tool
+assumes they are not present and does diagnostics only as a client.
+
+You may also specify config files explicitly with flags below, in which
+case you will receive an error if they are not found or invalid.
+
+    $ %[1]s --master-config=/etc/openshift/master.yaml
+
+Subcommands may be used to scope the troubleshooting to a particular
+component and are not limited to using config files; you can and should
+use the same flags that are actually set on the command line for that
+component to configure the diagnostic.
+
+    $ %[1]s node --master=https://master.openshift.example.com:8443 --cert-dir=...
+
+NOTE: This is an alpha version of diagnostics and will change significantly.
+NOTE: Global flags (from the 'options' subcommand) are ignored here but
+can be used with subcommands.
+`
+
+func NewCommandDiagnostics(name string, fullName string, out io.Writer) *cobra.Command {
+	opts := options.NewAllDiagnosticsOptions(out)
+	cmd := &cobra.Command{
+		Use:   name,
+		Short: "This utility helps you understand and troubleshoot OpenShift v3.",
+		Long:  fmt.Sprintf(longAllDescription, fullName),
+		Run: func(c *cobra.Command, args []string) {
+			opts.GlobalFlags = c.PersistentFlags()
+			run.Diagnose(opts)
+		},
+	}
+	cmd.SetOutput(out) // for output re: usage / help
+	opts.BindFlags(cmd.Flags(), options.NewAllDiagnosticsFlagInfos())
+	// Although we reuse DiagOptions across all commands, we do not want the flags buried
+	// in the "global" flags, so we add them locally at each command.
+	opts.DiagOptions.BindFlags(cmd.Flags(), options.NewDiagnosticsFlagInfos())
+
+	/*
+	   This command needs the client factory built in the "client" subcommand.
+	   Generating the factory adds flags to the "client" cmd, and we do not want
+	   to add those flags to this command (the only client option here is a config
+	   file). So the factory object from client cmd is reused for this command.
```

Comment:
Wait, what?  There's probably a better way.  I'll think about it during the review.